keymanapp/keyman
# fix(core): reset on frame keys 🙀

- Move context reset out of kmx and into the core
- use the context reset table
- reinstate tests in ldml that test for reset
- improve ldml test suite to support multiple steps in the 'embedded' test format

Fixes #10955
Fixes #10227

# User Testing 

Using Windows 10 or Windows 11.
Install the keyman build from this PR.

Use the Text Editor linked in this PR under test artifacts. or if on Windows 10 you can use Notepad.

## SUITE_KMX_PROCESSOR_NON_COMPLIANT:
For these tests make sure the application used on the operating system is a **non-compliant**. A app that is not able to give the context to Keyman.

#### GROUP_WIN_10: TextEditor
#### GROUP_WIN_11: TextEditor
#### GROUP_MAC:  Chrome Browser
#### GROUP_LINUX:  Chrome Brower (https://www.editpad.org/)

Install SIL Euro Latin

### TEST_CONTROL_1
1. Press <kbd>`</kbd> 
2. press <kbd>e</kbd>
Expected result:
è

### TEST_FRAME_KEY_RESET_MARKERS
This tests the markers are lost
1. Press <kbd>`</kbd> 
2. Press <kbd>-></kbd>    right arrow
3. Press <kbd>e</kbd>
Expected result: 
`e

### TEST_FRAME_KEY_RESET_NOMARKERS

1. Press <kbd>1</kbd> <kbd>/</kbd> <kbd>/</kbd>
2. Press <kbd>-></kbd>    right arrow
3. Press <kbd>2</kbd>
Expected result: 
1//2

### TEST_SCROLLLOCK_KEY_NO_RESET
Do not reset for scroll lock
1. Press <kbd>`</kbd>
2. Press <kbd>scroll lock</kbd> for Windows and Linux. For MacOS Press <kbd>fn</kbd>
3. Press <kbd>e</kbd>
Expected Result: 
è

## SUITE_KMX_PROCESSOR_COMPLIANT:

For these tests make sure the application used on the operating system is **compliant**. That is Keyman can request the context from the application.

#### GROUP_WIN_10: WordPad
#### GROUP_WIN_11: Wordpad
#### GROUP_MAC: TextEdit 
#### GROUP_LINUX: gedit

Install SIL Euro Latin


### TEST_CONTROL_1
1. Press <kbd>`</kbd> 
2. press <kbd>e</kbd>
Expected result:
è

### TEST_FRAME_KEY_RESET_MARKERS
This tests the markers are lost
1. Press <kbd>`</kbd> 
2. Press <kbd>-></kbd>    right arrow
3. Press <kbd>e</kbd>
Expected result: 
`e

### TEST_FRAME_KEY_RESET_NOMARKERS

1. Press <kbd>1</kbd> <kbd>/</kbd> <kbd>/</kbd>
2. Press <kbd>-></kbd>    right arrow
3. Press <kbd>2</kbd>
Expected result: 
½

### TEST_SCROLLLOCK_KEY_NO_RESET
Do not reset for scroll lock
1. Press <kbd>`</kbd>
2. Press <kbd>scroll lock</kbd> for Windows and Linux. For MacOS Press <kbd>fn</kbd>
3. Press <kbd>e</kbd>
Expected Result: 
è

## SUITE_LDML_PROCESSOR_NON_COMPLIANT:
LDML keyboard

### GROUP_WIN_10: TextEditor
### GROUP_WIN_11: TextEditor
### GROUP_MAC:  Chrome Browser
### GROUP_LINUX:  Chrome Brower

Install [contextreset.zip](https://github.com/keymanapp/keyman/files/14976065/contextreset.zip) ldml keyboard
### TEST_CONTROL_1
1. Press <kbd>^</kbd> 
2. Press <kbd>e</kbd>
Expected result:
ê

This was previously failing 

### TEST_FRAME_KEY_RESET_MARKERS
1. Press <kbd>^</kbd>
2. Press <kbd>-></kbd>    right arrow
3. Press <kbd>e</kbd>
Expected result: 
e

### TEST_FRAME_KEY_RESET_NO_MARKERS
1. Press <kbd>a</kbd>
2. Press <kbd>-></kbd>    right arrow
3. Press <kbd>Shift</kbd> + <kbd>8</kbd> 
Expected result: 
a*

### TEST_FRAME_SCROLL_LOCK_NO_RESET
1. Press <kbd>^</kbd>
2. Press <kbd>scroll lock</kbd> for Windows and Linux. For MacOS Press <kbd>fn</kbd>
3. Press <kbd>e</kbd>
Expected result: 
ê

### TEST_MODIFIER_TAP_NO_RESET
Do not reset for modifier tap
1. Press <kbd>^</kbd>
2. Press  and Release <kbd>ctrl</kbd>
3. Press <kbd>e</kbd>
Expected result:
ê

## SUITE_LDML_PROCESSOR_COMPLIANT:
LDML keyboard

### GROUP_WIN_10: WordPad
### GROUP_WIN_11: WordPad
### GROUP_MAC: TextEdit 
### GROUP_LINUX: gedit

Install
[contextreset.zip](https://github.com/keymanapp/keyman/files/14976065/contextreset.zip) ldml keyboard
### TEST_CONTROL_1
1. Press <kbd>^</kbd> 
2. Press <kbd>e</kbd>
Expected result:
ê

This was previously failing 

### TEST_FRAME_KEY_RESET_MARKERS
1. Press <kbd>^</kbd>
2. Press <kbd>-></kbd>    right arrow
3. Press <kbd>e</kbd>
Expected result: 
e

### TEST_FRAME_KEY_RESET_NO_MARKERS
1. Press <kbd>a</kbd>
2. Press <kbd>-></kbd>    right arrow
3. Press <kbd>Shift</kbd> + <kbd>8</kbd> 
Expected result: 
Star

### TEST_FRAME_SCROLL_LOCK_NO_RESET
1. Press <kbd>^</kbd>
2. Press <kbd>scroll lock</kbd> for Windows and Linux. For MacOS Press <kbd>fn</kbd>
3. Press <kbd>e</kbd>
Expected result: 
ê

### TEST_MODIFIER_TAP_NO_RESET
Do not reset for modifier tap
1. Press <kbd>^</kbd>
2. Press  and Release <kbd>ctrl</kbd>
3. Press <kbd>e</kbd>
Expected result:
ê
